### ISSUE TITLE
Introduce Opaque/Customized Operator

### ISSUE THREAD

**Issue by houseroad** [onnx/onnx#471]
ONNX cannot cover all types of operators. But if users still want to represent their models using onnx, we probably should allow them to do so.

The proposal is adding experimental `opaque`/`customized` operators. So backend support is optional, and users can still represent their models in ONNX format, do the things they want, such as applying optimization.

**Comment by ezyang** [comment]
@houseroad, isn't this what operators with non-default domain do?

**Comment by houseroad** [comment]
@ezyang I agree non-default domain can achieve the goal, but it requires the user to define their own op set. is it too much work just for some blackbox operator?

**Comment by ezyang** [comment]
I don't see why it is all that much work, you just pick a new name out of your operator namespace and just use it.

Perhaps the missing piece is that the validator needs to ignore operators that it doesn't understand. It seems useful, in this case, to have a "strict" and "relaxed" mode for the validator.

**Comment by rockindy** [comment]
@houseroad Do we have standard interface to add custom operators?
I want to add custom operators for SSH and YOLOv3 and hope those operators can be supported by shape inference.

**Comment by llhe** [comment]
Customized Op is very common in production. For example, as an inference framework creator (https://github.com/XiaoMi/mace), we have many customized Ops in Caffe/PyTorch and also Kaldi. This limitation makes it hard to get a clean solution with ONNX as intermediate representation and hinders the application of ONNX.

**Comment by prasanthpul** [comment]
We use custom ops for some of our models as well. ONNX supports this well thanks to the concept of domains. The ops in the default domain are well defined and generally should be implemented across tools. But you can have your own ops in your own domain. This also allows tools to examine a model and determine whether they can run it.

**Comment by llhe** [comment]
@prasanthpul Thanks, that's nice feature, I'll have a look.
